posthog/renderers.py

```diff
@@ -0,0 +1,46 @@
+import math
+from typing import Dict, Tuple
+from rest_framework.renderers import JSONRenderer
+
+CleaningMarker = bool | Dict[int, "CleaningMarker"]
+
+
+def clean_data_for_json(data) -> Tuple[CleaningMarker, CleaningMarker]:
+    """Replace NaNs and Infinities with None, in-place, marking which fields had to be scrubbed.
+    Return markers that should be set at the parent level, which is used for setting the markers recursively."""
+    if isinstance(data, float):
+        return (math.isnan(data), math.isinf(data))
+    if isinstance(data, list):
+        nan_map: Dict[int, CleaningMarker] = {}
+        inf_map: Dict[int, CleaningMarker] = {}
+        for index, item in enumerate(data):
+            nan, inf = clean_data_for_json(item)
+            if nan:
+                nan_map[index] = nan
+                if nan is True:
+                    data[index] = None
+            if inf:
+                inf_map[index] = inf
+                if inf is True:
+                    data[index] = None
+        return (nan_map, inf_map)
+    if isinstance(data, dict):
+        marker_payload: Dict[str, CleaningMarker] = {}
+        for key, value in data.items():
+            nan, inf = clean_data_for_json(value)
+            if nan:
+                marker_payload[f"{key}::nan!"] = nan
+                if nan is True:
+                    data[key] = None
+            if inf:
+                marker_payload[f"{key}::inf!"] = inf
```

Comment:
The point is to avoid changing the shape of data (i.e. no interfaces need to change in JS), that's why the nan/inf markers are peers of the bad key. We're not going to use these error keys anywhere in the code, but we're going to need them when a user asks why something isn't working and we find out there's a null – then it's important to know whether that's a "natural" null or a "fixed" NaN. Hence the "!" – it's just so that the problem stands out 😄

Reply:
Oh I see, so no frontend changes needed. I do feel like the `!` is kind of dirty/unnecessary but if you like it then I get your reasoning.

Reply:
Fair, I guess it's just not really helpful then. Removed